syntax = "proto3";

package distributed;

import "google/protobuf/timestamp.proto";

// Service for server-to-server communication
service DistributedService {
  // Leader election and heartbeat
  rpc RequestVote(VoteRequest) returns (VoteResponse);
  rpc SendHeartbeat(HeartbeatRequest) returns (HeartbeatResponse);
  
  // State replication
  rpc ReplicateCommand(CommandRequest) returns (CommandResponse);
  rpc SyncState(SyncRequest) returns (SyncResponse);
  
  // Data operations
  rpc CreateUser(UserRequest) returns (UserResponse);
  rpc AuthenticateUser(AuthRequest) returns (AuthResponse);
  rpc CreateDocument(DocRequest) returns (DocResponse);
  rpc UpdateDocument(DocUpdateRequest) returns (DocUpdateResponse);
  rpc DeleteDocument(DocDeleteRequest) returns (DocDeleteResponse);
  rpc AddUserToDocument(DocUserRequest) returns (DocUserResponse);
  rpc RemoveUserFromDocument(DocUserRequest) returns (DocUserResponse);
}

// Leader election messages
message VoteRequest {
  string server_id = 1;
  int64 term = 2;
  int64 last_log_index = 3;
  int64 last_log_term = 4;
}

message VoteResponse {
  string server_id = 1;
  int64 term = 2;
  bool vote_granted = 3;
}

message HeartbeatRequest {
  string leader_id = 1;
  int64 term = 2;
  int64 commit_index = 3;
  repeated LogEntry entries = 4;
}

message HeartbeatResponse {
  string server_id = 1;
  int64 term = 2;
  bool success = 3;
  int64 last_applied = 4;
}

// Log replication
message LogEntry {
  int64 term = 1;
  int64 index = 2;
  bytes command = 3;
  google.protobuf.Timestamp timestamp = 4;
}

message CommandRequest {
  string leader_id = 1;
  int64 term = 2;
  LogEntry entry = 3;
}

message CommandResponse {
  string server_id = 1;
  int64 term = 2;
  bool success = 3;
}

// State synchronization
message SyncRequest {
  string server_id = 1;
  int64 last_index = 2;
}

message SyncResponse {
  repeated LogEntry entries = 1;
  bool success = 2;
}

// Data operation messages
message UserRequest {
  string username = 1;
  string password = 2;
}

message UserResponse {
  bool success = 1;
  string message = 2;
}

message AuthRequest {
  string username = 1;
  string password = 2;
}

message AuthResponse {
  bool success = 1;
  string message = 2;
}

message DocRequest {
  string title = 1;
  string username = 2;
}

message DocResponse {
  bool success = 1;
  string message = 2;
  string document_id = 3;
}

message DocUpdateRequest {
  string document_id = 1;
  string username = 2;
  string content = 3;
  string base_content = 4;
  string title = 5;
  bool is_content_update = 6;
  bool is_title_update = 7;
}

message DocUpdateResponse {
  bool success = 1;
  string message = 2;
  string content = 3;
}

message DocDeleteRequest {
  string document_id = 1;
  string username = 2;
}

message DocDeleteResponse {
  bool success = 1;
  string message = 2;
}

message DocUserRequest {
  string document_id = 1;
  string username = 2;
  string requester = 3;
}

message DocUserResponse {
  bool success = 1;
  string message = 2;
}
